train_model.py

```diff
@@ -460,34 +497,39 @@ def get_params(self) -> dict:
     print(f"Average Test Loss: {avg_loss}")
 
     # Save hyperparameters and metrics in logs
-    writer.add_hparams(hparam_dict=model.get_params(),
-                       metric_dict={
-                           "Loss/train": avg_train_loss,
-                           "Loss/validation": avg_val_loss,
-                           "Loss/test": avg_loss})
+    writer.add_hparams(
+        hparam_dict=model.get_params(),
+        metric_dict={
+            "Loss/train": avg_train_loss,
+            "Loss/validation": avg_val_loss,
+            "Loss/test": avg_loss})
 
     # Close SummaryWriter
     writer.close()
 
     # Save completed model
-    weight_filename = f"training_completed.pt"
+    weight_filename = "training_completed.pt"
     torch.save(model.state_dict(), save_folder / weight_filename)
     print(f"Saved final weights as {weight_filename}")
 
+    # Save BPE tokenizer used
+    tokenizer.save_pretrained(save_directory=save_folder, filename_prefix="BPE")
```

Comment:
Please move this to the beginning before training begins. The vocabulary is needed, especially if the training crashes.

Reply:
Well said! This has been addressed in [54a977d](https://github.com/DRAGNLabs/301r_retnet/pull/15/commits/54a977dbb687efbeee52151b3e13cfbc3c2748cc).